### PR TITLE
Add support for Razer BlackWidow v3 Pro

### LINKED ISSUE #1279: Support for BlackWidow V3 Pro [1532:025c]
### Device Info

* Name: BlackWidow V3 Pro
* Product Number: RZ03-03530200-R3U1
* Model Number: RZ03-0353

### Device Information

<details>
  <summary> lsusb -d 1532: -v </summary>

  ```
Bus 001 Device 003: ID 1532:025c Razer USA, Ltd Razer BlackWidow V3 Pro
Couldn't open device, some information will be missing
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x025c 
  bcdDevice            1.00
  iManufacturer           1 
  iProduct                2 
  iSerial                 3 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0086
    bNumInterfaces          5
    bConfigurationValue     1
    iConfiguration          4 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              100mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     179
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      94
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        4
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     325
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x85  EP 5 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        3
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     348
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x84  EP 4 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
  ```
</details>
    
<details>
  <summary>ls -lah /dev/input/by-id/</summary>

  ```
total 0
drwxr-xr-x 2 root root 240 Oct  3 20:34 .
drwxr-xr-x 4 root root 800 Oct  3 20:34 ..
lrwxrwxrwx 1 root root   9 Oct  1 17:28 usb-Logitech_USB_Receiver-if02-event-kbd -> ../event6
lrwxrwxrwx 1 root root   9 Oct  1 17:28 usb-Logitech_USB_Receiver-if02-event-mouse -> ../event4
lrwxrwxrwx 1 root root   9 Oct  1 17:28 usb-Logitech_USB_Receiver-if02-mouse -> ../mouse2
lrwxrwxrwx 1 root root  10 Oct  3 20:34 usb-Razer_Razer_BlackWidow_V3_Pro_000000000000-event-if01 -> ../event27
lrwxrwxrwx 1 root root  10 Oct  3 20:34 usb-Razer_Razer_BlackWidow_V3_Pro_000000000000-event-if03 -> ../event29
lrwxrwxrwx 1 root root  10 Oct  3 20:34 usb-Razer_Razer_BlackWidow_V3_Pro_000000000000-event-kbd -> ../event23
lrwxrwxrwx 1 root root  10 Oct  3 20:34 usb-Razer_Razer_BlackWidow_V3_Pro_000000000000-if01-event-kbd -> ../event24
lrwxrwxrwx 1 root root  10 Oct  3 20:34 usb-Razer_Razer_BlackWidow_V3_Pro_000000000000-if02-event-mouse -> ../event28
lrwxrwxrwx 1 root root   9 Oct  3 20:34 usb-Razer_Razer_BlackWidow_V3_Pro_000000000000-if02-mouse -> ../mouse4
lrwxrwxrwx 1 root root   9 Oct  1 17:28 usb-SteelSeries_Arctis_Pro_Wireless-event-if05 -> ../event7
  ```
</details>

### Packet Captures

<Please see here as we probably need packet captures>
<https://github.com/openrazer/openrazer/wiki/Reverse-Engineering-USB-Protocol>

### Synapse Screenshots
<Screenshots of all available options as seen on Razer Synapse>
(Coming soon)

### LINKED ISSUE #1619: Support for BlackWidow V3 Pro [1532:025a]
### Device Info

* Name: Razer BlackWidow V3 Pro
* Product Number: RZ03-03530100
* Model Number: RZ03-0353

### Device Information
```
Please post the output of
    lsusb -d 1532: -v
```
 lsusb -d 1532:025a -v 

Bus 001 Device 006: ID 1532:025a Razer USA, Ltd Razer BlackWidow V3 Pro
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x025a 
  bcdDevice            2.00
  iManufacturer           1 Razer
  iProduct                2 Razer BlackWidow V3 Pro
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x006d
    bNumInterfaces          4
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     177
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      90
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        3
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     348
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x84  EP 4 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
Device Status:     0x0000
  (Bus Powered)

```
Please post the output of
    ls -lah /dev/input/by-id/
```
lrwxrwxrwx 1 root root  10 Sep 26 18:12 usb-Razer_Razer_BlackWidow_V3_Pro-event-if01 -> ../event24
lrwxrwxrwx 1 root root  10 Sep 26 18:12 usb-Razer_Razer_BlackWidow_V3_Pro-event-if03 -> ../event26
lrwxrwxrwx 1 root root  10 Sep 26 18:12 usb-Razer_Razer_BlackWidow_V3_Pro-event-kbd -> ../event22
lrwxrwxrwx 1 root root  10 Sep 26 18:12 usb-Razer_Razer_BlackWidow_V3_Pro-if01-event-kbd -> ../event23
lrwxrwxrwx 1 root root  10 Sep 26 18:12 usb-Razer_Razer_BlackWidow_V3_Pro-if02-event-mouse -> ../event25
lrwxrwxrwx 1 root root   9 Sep 26 18:12 usb-Razer_Razer_BlackWidow_V3_Pro-if02-mouse -> ../mouse1

### Packet Captures

<Please see here as we probably need packet captures>
<https://github.com/openrazer/openrazer/wiki/Reverse-Engineering-USB-Protocol>

### Synapse Screenshots
<Screenshots of all available options as seen on Razer Synapse>

### PR DESCRIPTION
**Pull request by faridgh1991**
Fixes #1279 
Building from source instructions can be found here: https://github.com/openrazer/openrazer/wiki/Building-a-package

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -69,6 +69,9 @@ The devices below are fully feature supported by OpenRazer, which means all avai
 | Razer Blade 15 (Early 2020) Base              |  1532:0255  |
 | Razer Huntsman Mini                           |  1532:0257  |
 | Razer Blade Stealth (Late 2020)               |  1532:0259  |
+| Razer BlackWidow V3 PRO (USB)                 |  1532:025A  |
+| Razer BlackWidow V3 PRO (Bluetooth)           |  1532:025B  |
+| Razer BlackWidow V3 PRO (WIFI)                |  1532:025C  |
```
Comment (z3ntu) [inline README.md]: From looking at the product page it doesn't look like it's using Wifi (which would be weird anyways). The product page says this, so maybe use this wording?
* Wired
* 2.4 Ghz Wireless
* Bluetooth

Also please fix the product name to have the correct capitalization (in all files): `Razer BlackWidow V3 Pro`

File: daemon/openrazer_daemon/hardware/keyboards.py [hunk 2]
```diff
@@ -344,6 +345,52 @@ class RazerBlackWidowUltimate2013(_MacroKeyboard):
...
+    """
+    Class for the Razer BlackWidow v3 PRO bluetooth
+    """
+    USB_PID = 0x025B
+    METHODS = RazerBlackWidowv3ProUSB.METHODS + ['set_charge_effect', 'set_charge_colour']
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self._battery_manager = _BatteryManager(self, self._device_number, 'Razer BlackWidow v3 PRO bluetooth')
```
Comment (z3ntu) [inline daemon/openrazer_daemon/hardware/keyboards.py]: The BatteryManager needs the `getBattery` function to work (as seen in the CI failure), so you probably want to add these methods to the METHODS array:
```
               'get_battery', 'is_charging', 'set_idle_time', 'set_low_battery_threshold',
```
You need to add the needed kernel parts as well and then validate that the methods are working

File: driver/razerkbd_driver.h [hunk 3]
```diff
@@ -36,6 +36,9 @@
 #define USB_DEVICE_ID_RAZER_BLACKWIDOW_X_CHROMA 0x0216
 #define USB_DEVICE_ID_RAZER_BLACKWIDOW_X_ULTIMATE 0x0217
 #define USB_DEVICE_ID_RAZER_BLACKWIDOW_X_CHROMA_TE 0x021A
+#define USB_DEVICE_ID_RAZER_BLACKWIDOW_V3_PRO_USB 0x025A
+#define USB_DEVICE_ID_RAZER_BLACKWIDOW_V3_PRO_BLUETOOTH 0x025B
+#define USB_DEVICE_ID_RAZER_BLACKWIDOW_V3_PRO_WIFI 0x025C
```
Comment (z3ntu) [inline driver/razerkbd_driver.h]: Please keep these entries sorted by PID, so put these after `USB_DEVICE_ID_RAZER_BLADE_STEALTH_LATE_2020`

**Review by z3ntu — CHANGES_REQUESTED** [review]
Thanks for the PR! I've added some comments below

File: README.md (hunk 1 above)
Comment (z3ntu) [reply README.md]: This still seems to be an open issue

> Also please fix the product name to have the correct capitalization (in all files): Razer BlackWidow V3 Pro

File: daemon/openrazer_daemon/hardware/keyboards.py (hunk 2 above)
Comment (faridgh1991) [reply daemon/openrazer_daemon/hardware/keyboards.py]: > The BatteryManager needs the `getBattery` function to work (as seen in the CI failure), so you probably want to add these methods to the METHODS array:
> 
> ```
>                'get_battery', 'is_charging', 'set_idle_time', 'set_low_battery_threshold',
> ```
> 
> You need to add the needed kernel parts as well and then validate that the methods are working

@z3ntu 
I have problems with battery methods, can you help me?
```Traceback (most recent call last):
  File "/usr/lib/python3.9/threading.py", line 954, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3/dist-packages/openrazer_daemon/misc/battery_notifier.py", line 105, in run
    self.notify_battery()
  File "/usr/lib/python3/dist-packages/openrazer_daemon/misc/battery_notifier.py", line 75, in notify_battery
    battery_level = self._get_battery_func()
  File "/usr/lib/python3/dist-packages/openrazer_daemon/dbus_services/dbus_methods/mamba.py", line 15, in get_battery
    with open(driver_path, 'r') as driver_file:
FileNotFoundError: [Errno 2] No such file or directory: '/sys/devices/pci0000:00/0000:00:14.0/usb1/1-4/1-4:1.2/0003:1532:025A.0030/charge_level'
```

Comment (z3ntu) [reply daemon/openrazer_daemon/hardware/keyboards.py]: According to the driver this should exist. Are you sure you have the latest version loaded?

**Comment by faridgh1991** [comment]
battery functions work now.
in wireless mode device recognized as "Razer BlackWidow V3 Pro(Wireless)" and an unknown serial (RazerKeyboard UNKWN00000000XXXX) and no function work.
in Bluetooth mode device not recognized by daemon.

**Comment by z3ntu** [comment]
For the wireless functionality you can try adjusting the timeout in `razer_get_report` - compare with razermouse_driver.c and take RAZER_NEW_MOUSE_RECEIVER_WAIT_MIN_US/RAZER_NEW_MOUSE_RECEIVER_WAIT_MAX_US for this device

**Comment by geodox** [comment]
I have cloned the repo, checked out this branch, rebased (to update the local files to these changes), built the debs and installed them and I still can't get my BW V3 Pro to work in Wired mode. Can someone provide me instructions to get wired mode to be accepted in by openrazer (if it's working based on the comments about battery mode)?

I'm using polychromatic and the troubleshooter says it's still incompatible.

Edit: I did manage to get it working after several restarts.

**Comment by EduardTheFirst** [comment]
> I have cloned the repo, checked out this branch, rebased (to update the local files to these changes), built the debs and installed them and I still can't get my BW V3 Pro to work in Wired mode. Can someone provide me instructions to get wired mode to be accepted in by openrazer (if it's working based on the comments about battery mode)?
> 
> I'm using polychromatic and the troubleshooter says it's still incompatible.

me too

**Comment by EduardTheFirst** [comment]
any update on this pull?
if i can contribute please tell me how

**Comment by samoht9277** [comment]
Second this. Any way to get some sort of lighting effects or battery info?

**Comment by geodox** [comment]
> Second this. Any way to get some sort of lighting effects or battery info?

Not at the moment seemingly, besides plugging the device into a windows machine and configuring it from there. Have you got the device recognized using this PR?

**Comment by samoht9277** [comment]
> Have you got the device recognized using this PR?

How exactly can i do this? compiling the driver, and installing a front-end?

**Comment by geodox** [comment]
After several restarts, this build of open-razer finally loaded and recognized my BW V3 Pro. Everything that I have tried works in wired mode. Battery is recognized (even in wired mode). Will post back after I try bluetooth and wireless modes.
image

**Comment by geodox** [comment]
> How exactly can i do this? compiling the driver, and installing a front-end?

The steps for building were posted above. Here's the link again. https://github.com/openrazer/openrazer/wiki/Building-a-package
Clone the authors repo, rebase to this branch, build and install the debs (you may have to remove previously installed versions of open-razer for it to work), install a frontend and finally restart (more than once if neccessary).

**Comment by samoht9277** [comment]
@GeoDoX finally could install the driver, but I cloned [this](https://github.com/faridgh1991/openrazer) repo, built the packages, installed polychromatic and added `openrazer-driver-dkms-git` provided by the compilation process.

The only thing left now is to clone this repo and rebase [this](https://github.com/faridgh1991/openrazer) to get the latest `openrazer` version. Do you know how may i do that?

proof 
image

**Comment by geodox** [comment]
@samoht9277 It's up to the author to do that, unless you want to clone this into your own repo and merge the changes yourself. Either way, when this get's merged into official, you'll have to uninstall this version and install the latest official. Up to you if you think it's worth the version bump.

File: pylib/openrazer/_fake_driver/razerblackwidowv3problutooth.cfg [hunk 4]
```diff
@@ -0,0 +1,26 @@
+[device]
+dir_name = 0003:1532:025B.0001
+name = Razer BlackWidow V3 Pro(Blutooth)
```
Comment (GeoDoX) [inline pylib/openrazer/_fake_driver/razerblackwidowv3problutooth.cfg:3]: Is there a chance the daemon isn't picking up bluetooth mode due to this file (as well as the contents) reference bluetooth as 'blutooth'?

**Comment by z3ntu** [comment]
Closing in favor of #1622 and #1623 , please especially test #1622!

**Pull request closed without merging**